File: Javascript/20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const startParent = new Map([["(", ")"], ["{", "}"], ["[", "]"]]);
    const endParent = new Map([[")", "("], ["}", "{"], ["]", "["]]);
    const splitedS = s.split('');

    let stack = [];
    for (i = 0; i < splitedS.length; i++) {
        const pair = startParent.get(splitedS[i]);
        if (pair !== undefined) {
            stack.push(splitedS[i]);
            continue;
        }

        if (stack.length === 0) {
            return false;
        }

        const popped = stack.pop();
        if (endParent.get(splitedS[i]) !== popped) {
            return false;
        }
    }

    return stack.length === 0;
};
// @lc code=end

console.log(isValid("()"));

File: Ruby/20.valid-parentheses.rb
#
# @lc app=leetcode id=20 lang=ruby
#
# [20] Valid Parentheses
#

# @lc code=start
# @param {String} s
# @return {Boolean}
def is_valid(s)
    map = { "(" => ")", "{" => "}", "[" => "]" }
    stack = []

    for i in 0...s.size
        if s[i] == "(" || s[i] == "{" || s[i] == "["
            stack.append s[i]
            next
        end

        if s[i] != map[stack.pop]
            return false
        end
    end

    return stack.size == 0
end
# @lc code=end
